### ISSUE TITLE
[amdgpu][mc][fuzzer-generated] Out of range scalar-register fields assert in DecodeSReg callback

### ISSUE THREAD

**Issue by kuhar** [llvm/llvm-project#214915]
This is a fuzzer-generated report, assisted by codex. It was found by passing mutated gfx1250 instruction bytes through the LLVM disassembler.

## Version

Tested at current LLVM `main` commit
`bc495da845041056e9a5892816d7fc214e371716` (2026-08-07) with an
assertions-enabled `RelWithDebInfo` build of `llvm-mc`.

## Minimized reproducers

The three minimized byte streams and complete commands are:

```sh
# DecodeSReg_32RegisterClass
printf '%s\n' '0xff 0x40 0x80 0xbe' |
  bin/llvm-mc \
    --triple=amdgcn-amd-amdhsa --mcpu=gfx1250 \
    --disassemble --show-encoding

# DecodeSReg_32_XEXECRegisterClass
printf '%s\n' '0xfa 0xc1 0x80 0xd6 0xff 0x00 0x00 0x00' |
  bin/llvm-mc \
    --triple=amdgcn-amd-amdhsa --mcpu=gfx1250 \
    --disassemble --show-encoding

# DecodeSReg_32_XM0RegisterClass
printf '%s\n' '0xfe 0x04 0x00 0x7f' |
  bin/llvm-mc \
    --triple=amdgcn-amd-amdhsa --mcpu=gfx1250 \
    --disassemble --show-encoding
```

The process exits with status 134. The callback name changes, but every failure
reports:

```text
Assertion `Imm < (1 << 7) && "7" "-bit encoding"' failed.
```

These selectors do not name valid registers for the decoded operands. LLVM
should report an invalid instruction encoding without aborting.

## Root cause

Note: this is clanker-generated.

`DECODE_SrcOp` asserts that the extracted immediate fits its declared encoding
size. The three generated callbacks are instantiated through
`DECODE_OPERAND_SREG_7`, so malformed instruction fields of 128 or greater
reach the same seven-bit assertion. Decoder callbacks operate on untrusted
machine-code fields; an out-of-range field is a decode failure, not an internal
invariant.

The assertion was added by commit
`b0c1a45ba5e5f9f04f5e208a6efb120dbb903758` in 2023. The preceding callbacks
did not contain this unconditional bounds assertion.

The shared decoder macro can bounds-check the extracted value and return
`MCDisassembler::Fail` before calling `decodeSrcOp`, preserving assertions for
true internal invariants while treating malformed fields as ordinary decode
failures.

## Plausibility and deduplication

These are malformed raw instruction words. Arbitrary fields are expected at a
disassembler boundary, and the generated decoder should return
`MCDisassembler::Fail` for them.

The final gfx1250 replay contained two generic, five XEXEC, and one XM0
failure. They share the macro, assertion, and required handling, so they are
one report rather than three. Cross-target checks on current `main` also show
the XM0 input asserting on gfx900, gfx90a, gfx1030, gfx1100, and gfx1200; the
other two variants reproduce on the newer targets whose tables select those
callbacks. The defect is therefore AMDGPU-wide rather than specific to
gfx1250.
